### PR TITLE
MEPTS-269: Changes in common queries for IM_ER

### PR DESCRIPTION
**Pull request by ynurmahomed**

### CODE REVIEW

**Comment by nthfloor** [comment]
@ynurmahomed @KelvinWanyama there are merge conflicts here. Please review and resolve them

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/generic/InitialArtStartDateCalculation.java [hunk 1]
```diff
@@ -181,6 +181,7 @@ public static Date getArtStartDate(Integer patientId, CalculationResultMap artSt
 
   private CalculationResultMap getMastercardDrugPickupMap(
       Collection<Integer> cohort, Location location, PatientCalculationContext context) {
+    //    TODO: also handle Levantou ARV (concept id 23865) = SIm (1065)
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/generic/InitialArtStartDateCalculation.java]: @ynurmahomed is this meant to remain as TODO? 

In the spec this is highlighted as green, which I understood means it should be added...

**Comment by ynurmahomed** [comment]
@nthfloor fixed.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java [hunk 2]
```diff
@@ -204,27 +205,123 @@ public CohortDefinition getPatientsBasedOnPatientStatesBeforeDate(int program, i
...
     cd.setName("Get deceased patients based on patient states and person object");
-    cd.addParameter(new Parameter("startDate", "Start Date", Date.class));
-    cd.addParameter(new Parameter("endDate", "End Date", Date.class));
+    cd.addParameter(new Parameter("onOrBefore", "End Date", Date.class));
     cd.addParameter(new Parameter("location", "Location", Location.class));
+    String sql =
+        "SELECT patient_id "
+            + "FROM   (SELECT patient_id, "
+            + "               Max(death_date) death_date "
+            + "        FROM   (SELECT p.patient_id, "
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:216]: There doesn't seem to be a closing bracket for this one...

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java [hunk 3]
```diff
@@ -204,27 +205,123 @@ public CohortDefinition getPatientsBasedOnPatientStatesBeforeDate(int program, i
...
+            + "                       AND pp.location_id = :location "
+            + "                       AND ps.voided = 0 "
+            + "                       AND ps.state = %d "
+            + "                       AND ps.start_date <= :onOrBefore "
+            + "                       AND ps.end_date IS NULL "
+            + "                UNION "
+            + "                SELECT p.person_id, "
+            + "                       p.death_date "
+            + "                FROM   person p "
+            + "                WHERE  p.voided = 0"
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:235]: Should there be a condition here for `death_date <=:endDate` ?

**Review by nthfloor — CHANGES_REQUESTED** [review]
@ynurmahomed I've reviewed the death query: a couple things to pls look at

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/PregnantQueries.java [hunk 4]
```diff
@@ -37,7 +39,7 @@ public static String getPregnantWhileOnArt(
         + adultInitailEncounter
         + ","
         + adultSegEncounter
-        + ") and e.encounter_datetime between :startDate and :endDate and e.location_id=:location and pe.gender='F' "
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/PregnantQueries.java]: @ynurmahomed seems this change was reverted with the last TxNew PR

We need to include the filter for females here. Pls re-add it.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 2 above)
Comment (nthfloor) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java]: ah I got confused here, actually I see it on 278. Sorry got confused there :man_facepalming:

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java [hunk 5]
```diff
@@ -191,6 +197,175 @@ public CohortDefinition getPatientsInArtWhoAbandoned() {
...
+            + "               AND e.voided = 0 "
+            + "               AND o.voided = 0 "
+            + "               AND o.concept_id = %s "
+            + "               AND o.value_datetime IS NOT NULL "
+            + "               AND e.encounter_type = %s) most_recent "
+            + "WHERE  most_recent.value_datetime = (SELECT Max(value_datetime) "
+            + "                                     FROM   obs "
+            + "                                     WHERE  encounter_id = "
+            + "                                            most_recent.encounter_id) "
+            + "       AND most_recent.value_datetime < Date_add(:onOrBefore, interval 30 day) "
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: @ynurmahomed so the specs says that the encounter date should be 30 days before the EndDate

"And add 30 days and this date should be < endDate"

But here we are adding 30days to the enddate - is this correct?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java [hunk 6]
```diff
@@ -191,6 +197,175 @@ public CohortDefinition getPatientsInArtWhoAbandoned() {
...
+        "SELECT p.patient_id "
+            + "FROM   patient p "
+            + "       inner join encounter e "
+            + "               ON e.patient_id = p.patient_id "
+            + "       inner join obs o "
+            + "               ON o.encounter_id = e.encounter_id "
+            + "WHERE  p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.voided = 0 "
+            + "       AND e.encounter_type NOT IN ( %s, %s, %s, %s ) "
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: @ynurmahomed this query will not have the same effect of filtering out patients without these encounters.

For example in cases where patients have other encounters outside of this list, they will still be excluded. This is the same issue that @cfaife had on TxCurr. Pls speak to him or we can have a call if you have any doubts about this

Comment (nthfloor) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: wait sorry I am getting confused, this is not exactly the same scenario as txCurr, pls ignore :)

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java [hunk 7]
```diff
@@ -191,6 +197,175 @@ public CohortDefinition getPatientsInArtWhoAbandoned() {
...
+            + "       inner join obs o "
+            + "               ON o.encounter_id = e.encounter_id "
+            + "WHERE  p.voided = 0 "
+            + "       AND e.voided = 0 "
+            + "       AND o.voided = 0 "
+            + "       AND e.encounter_datetime = (SELECT Max(encounter_datetime) "
+            + "                                   FROM   encounter "
+            + "                                   WHERE  patient_id = p.patient_id) "
+            + "       AND e.encounter_type IN ( %s, %s, %s ) "
+            + "       AND ( o.concept_id NOT IN ( %s, %s ) "
```
Comment (nthfloor) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: @ynurmahomed the specs says: "...or its contents are null"

shouldn't we rather be checking the obs value_coded or value_numeric is null?

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 6 above)
Comment (nthfloor) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: So for example if a patient has encounters that are not in this list, then this query will pick them, even though they have encounters that are in fact in this list

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 2 above)
Comment (nthfloor) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:216]: Sorry I got a bit confused here, pls ignore :man_facepalming: I see the matching closing bracket on line 278

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/generic/InitialArtStartDateCalculation.java (hunk 1 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/generic/InitialArtStartDateCalculation.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java (hunk 3 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/GenericCohortQueries.java:235]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/PregnantQueries.java (hunk 4 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/PregnantQueries.java]: done

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 5 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: This was supposed to be re-used from TX_CURR I forgot to remove it. Removed it now.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 6 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: Re-using from TX_CURR.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 7 above)
Comment (ynurmahomed) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: Re-using from TX_CURR.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java (hunk 5 above)
Comment (nthfloor) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/HivCohortQueries.java]: Hmm ok but it is in the specs - did Pinki say we should just ignore?

**Review by nthfloor — APPROVED** [review]

**Pull request merged**